design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,486 @@
+# Host IP pooling with NNCPs.
+
+## Summary
+
+Provide a mechanism to assign static IP configuration for secondary interfaces
+for cluster nodes, supporting cluster scaling up and down.
+
+## Motivation
+
+Using DHCP as the IP assignment mechanism has some drawbacks:
+- Introduce latencies on node restart.
+- It presents a single point of failure in case the DHCP server is failing
+- It depends on network setup that can be complex.
+
+To overcome the above-mentioned drawbacks, statically configured interfaces taking the address from a
+dynamic host IP pool is a solution. Since the interfaces are static there is no
+reconfiguration after some time or at node restart.
```

Comment:
When we hardcode the ip address statically per node, that's there until the user change it, this should work similar to that, we are not implementing DHCP lease.

Reply:
We need the reference to the pool at NNCE so the handler is able to reconstruct the offset at `NodeIPPool` if needed.

Reply:
But in case we want the pool to forget the IP, we don't want it to be reconstructed.

Reply:
The user IP will be forget by removing it from NNCE ippools at status, so it will not be bringed back.

Reply:
Is doable technically, but maybe is better to implemnent the mechanism with a annotation. 

```yaml
annotations:
  nmstate.io/reset-ipool: traffic1
```

We  can use this at NNCP (it will all nodes affected by it) or at NNCE (it will affect only the NNCE node).

Reply:
It is a good idea, which I think makes it more API driven, and less prone to errors.

Don't want to get in your way though, you can start small, and implement this after the MVP.

Just for future reference, this is how I see it happening (sure you think the same):
  - user would add the annotation
  - the controller would reconcile the enactments across all the cluster (not getting into detail of if each local agent resets its part of the pool or not)
  - once that's done, the controller removes the annotation.

If this happens, it would be interesting to throw an event / add a condition. 

I would favor the condition since that lives forever, while the event is deleted after a short while.